versions/development/SPEC.md

```diff
@@ -1049,6 +1057,43 @@ output {
 }
 ```
 
+#### Directory Outputs
+
+Directories can be declared as outputs just like files. For example:
+```
+output {
+  Directory d = "created/directory/"
+}
+```
+
+Notes:
+
+ - When a directory is output, the contents (including subdirectories) are considered part of the output but the path is not. 
+ - When a directory is output, its contents *do not overwrite* any pre-existing directory content. For example:
```

Comment:
Similar behaviour is not specified for Files, and cromwell uses that to allow different localisation strategies; soft-link, hard-link, copy. Depending on the cromwell settings, modifying an input to a task may modify the original File, and an Output File could be the same as an Input (hard link to the same data).

I am interested in the reasoning for Directories being different to Files.

Reply:
The entire execution model of WDL relies on all values being immutable one created. That assumption probably should be explicit in the spec if it isn't already (for *all* value types) - it's **the** core, underlying assumption used for task scheduling and cloud portability of the same WDL to different backends.

The localization strategies in Cromwell should really be thought of as an optimization to allow you to avoid the file-copying overheads, and *not* as a way to circumvent the immutable value assumptions of WDL.

Reply:
I think we agree, either; this should also be stated explicitly for Files, it should be stated explicitly once as a global, general rule for WDL, or it should not be stated here alone as that implies it is only for Directory.

Reply:
IMO no two input / output directories (except for the case of returning an unchanged input directory) should really ever be the same Object in the file system that is storing the inputs /outputs. 

Alot of people like the idea of mounting a single directory and making modifications to it and then those modifications magically become present in another jobs workspace. While I can see the draw from this (especially when working in a local computing environment) it is both fragile, and destined to cause more problems then its worth. Once A directory is created it 100% should be an immutable copy. Any modificaitons would result in a NEW directory

Reply:
@patmagee I do not think you should force it on Cromwell level, better just put warnings. Sometimes bioinformatic tools are so buggy that you have to do very crazy tricks to make them work

Reply:
@antonkulaga We're not discussing the Cromwell level, this is the WDL specification

Reply:
@antonkulaga putting wgat implementers would need to do aside to get this to work, I still think directories MUST be immutable constructs. Once a directory is created there should be no change that does not result in a new directory